## Ensemble Techniques-

Ensemble techniques are those techniques which use the predictions of various other predictors(classifiers or regressors) and then using voting methods, does its own predictions.
In other words, it will use the predictions given by the classifiers or regressors and then will vote that which of the predictions have occured more no. of times. It will use that prediction as its own prediction.

Ensemble methods work best when the predictors used are independent of each other.

WHY ENSEMBLE TECHNIQUES:
We use ensemble techniques at the end of the the project just to check if the models we built can give a better accuracy when combined together.

### 1. Voting Classifier

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

In [4]:
from sklearn.ensemble import VotingClassifier

In [5]:
from sklearn.metrics import roc_auc_score

In [6]:
data=pd.read_csv('diabetes.csv')

In [7]:
l=['Glucose','BloodPressure','SkinThickness','Insulin','BMI']
for i in l:
    data[i]=data[i].replace(0,np.mean(data[i]))

In [8]:
x=data.iloc[:,:8]
y=data['Outcome']

In [9]:
data

Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin   BMI  \
0              6    148.0           72.0      35.000000   79.799479  33.6   
1              1     85.0           66.0      29.000000   79.799479  26.6   
2              8    183.0           64.0      20.536458   79.799479  23.3   
3              1     89.0           66.0      23.000000   94.000000  28.1   
4              0    137.0           40.0      35.000000  168.000000  43.1   
..           ...      ...            ...            ...         ...   ...   
763           10    101.0           76.0      48.000000  180.000000  32.9   
764            2    122.0           70.0      27.000000   79.799479  36.8   
765            5    121.0           72.0      23.000000  112.000000  26.2   
766            1    126.0           60.0      20.536458   79.799479  30.1   
767            1     93.0           70.0      31.000000   79.799479  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [12]:
from sklearn.model_selection import train_test_split

In [11]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.30,random_state=0)

In [13]:
## We have to first build the pridictors which will contibute to voting clasifiers:

rf=RandomForestClassifier(random_state=0)
svm=SVC(random_state=0)
lr=LogisticRegression(random_state=0)
dt=DecisionTreeClassifier(random_state=0)

#### Voting Classifier has to different voting ways :
    1. Hard Voting
    2. Soft Voting
    

Hard Voting : When the classifier picks the majority vote of the predictions, it is hard voting.

In [22]:
vc=VotingClassifier(estimators=[('lr', lr), ('rf', rf), ('svm', svm),('dt',dt)],voting='hard')

In [23]:
vc.fit(x_train,y_train)

C:\Users\Shweta Chopra\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


VotingClassifier(estimators=[('lr',
                              LogisticRegression(C=1.0, class_weight=None,
                                                 dual=False, fit_intercept=True,
                                                 intercept_scaling=1,
                                                 l1_ratio=None, max_iter=100,
                                                 multi_class='auto',
                                                 n_jobs=None, penalty='l2',
                                                 random_state=0, solver='lbfgs',
                                                 tol=0.0001, verbose=0,
                                                 warm_start=False)),
                             ('rf',
                              RandomForestClassifier(bootstrap=True,
                                                     ccp_alpha=0.0,
                                                     class_weight=None,
                                                

In [24]:
y_ppred=vc.predict(x_test)

In [25]:
roc_auc_score(y_test,y_ppred) #accuracy given by voting classifier

0.6946978825959718

In [26]:
models=[rf,svm,lr,dt]
for i in models:
    i.fit(x_train,y_train)
    y_pred=i.predict(x_test)
    c=roc_auc_score(y_test,y_pred)
    print(c)

0.7336030297813738
0.6616887588225168
0.7026166293682218
0.724823549664314


C:\Users\Shweta Chopra\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### here the individual classifiers are giving better accuracies but they are instances when the voting classifier outcomes the predictors

Soft Voting:
If all classifiers are able to estimate class probabilities (i.e. they have a predict_proba() method), then you can tell Scikit-Learn to predict the class with the highest class probability, averaged over all the individual classifiers.

If any of the predictor is not any  predict_proba() method the we will set its probability hyperparameter as true.
Everything is done below.

In [28]:
vc1=VotingClassifier(estimators=[('rf',rf),('svm',svm),('dt',dt),('lr',lr)],voting='soft')

In [29]:
vc1.fit(x_train,y_train)

C:\Users\Shweta Chopra\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


VotingClassifier(estimators=[('rf',
                              RandomForestClassifier(bootstrap=True,
                                                     ccp_alpha=0.0,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     max_samples=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
        

In [30]:
y_pred1=vc1.predict(x_test)

AttributeError: predict_proba is not available when  probability=False

In [33]:
#HERE we can see that the svc predictor is not having the predict_proba method so we will so the hyperparameter correction
svm1=SVC(random_state=0,probability=True)

In [34]:
vc1=VotingClassifier(estimators=[('rf',rf),('svm1',svm1),('dt',dt),('lr',lr)],voting='soft')

In [35]:
vc1.fit(x_train,y_train)

C:\Users\Shweta Chopra\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


VotingClassifier(estimators=[('rf',
                              RandomForestClassifier(bootstrap=True,
                                                     ccp_alpha=0.0,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     max_samples=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
        

In [36]:
y_pred1=vc1.predict(x_test)

In [37]:
roc_auc_score(y_test,y_pred1) 

0.7467292132897229

Soft voting(74%) is better than the hard voting(69.5%) because it gives more weight to highly confident votes.

In [38]:
models=[rf,svm1,lr,dt]
for i in models:
    i.fit(x_train,y_train)
    y_pred=i.predict(x_test)
    c=roc_auc_score(y_test,y_pred)
    print(c) #accuracy of the predictpr

0.7336030297813738
0.6616887588225168
0.7026166293682218
0.724823549664314


C:\Users\Shweta Chopra\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
